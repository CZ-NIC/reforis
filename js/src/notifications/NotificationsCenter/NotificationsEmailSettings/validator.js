/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */


import {validateMultipleEmails} from '../../../forisForm/validation';

export default function validator(formData) {
    let errors = {};
    if (!formData.enabled)
        return undefined;

    errors.common = commonValidator(formData.common);
    if (formData.smtp_type === 'turris')
        errors.smtp_turris = smtpTurrisValidator(formData.smtp_turris);
    else if (formData.smtp_type === 'custom')
        errors.smtp_custom = smtpCustomValidator(formData.smtp_custom);
    return JSON.stringify(errors) !== '{}' ? errors : undefined;
}


function commonValidator(formData) {
    if (formData.to === '')
        return {to: _('Can\'t be empty.')};

    const toError = validateMultipleEmails(formData.to);
    return toError ? {to: toError} : undefined
}


const SENDER_NAME_RE = /^[0-9a-zA-Z_\\.-]+$/;

function smtpTurrisValidator(formData) {
    if (formData.sender_name === '')
        return {sender_name: _('Sender\'s name can\'t be empty.')};

    return !SENDER_NAME_RE.test(formData.sender_name) ?
        {sender_name: _("Sender's name can contain only alphanumeric characters, dots and underscores.")}
        : undefined
}

function smtpCustomValidator(formData) {
    if (isNaN(parseInt(formData.port)))
        return {port: _('Port is an number.')};
    if (formData.port > 65535)
        return {port: _('Maximum port is 65535.')};
    if (formData.port < 1)
        return {port: _('Minimum port is 1.')};
    return undefined
}
