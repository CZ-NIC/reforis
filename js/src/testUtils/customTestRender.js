/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

/* eslint import/export: "off" */

import React from 'react';
import {UIDReset} from 'react-uid';
import {StaticRouter} from 'react-router';
import {render} from '@testing-library/react'

function Wrapper({children}) {
    return <StaticRouter>
        <UIDReset>
            {children}
        </UIDReset>
    </StaticRouter>
}

const customTestRender = (ui, options) => render(ui, {wrapper: Wrapper, ...options});

// re-export everything
export * from '@testing-library/react'

// override render method
export {customTestRender as render}
