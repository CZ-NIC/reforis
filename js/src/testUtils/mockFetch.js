/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

export default function mockFetch(data) {
    return jest.fn(() =>
        new Promise((resolve, reject) => {
            resolve({
                ok: true,
                json: () => {
                    return data ? data : {};
                },
            });
        })
    );
}
