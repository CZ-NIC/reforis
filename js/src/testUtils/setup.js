/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import mockAxios from "jest-mock-axios";
import moment from "moment-timezone";

// Setup axios cleanup
global.afterEach(() => {
    mockAxios.reset();
});

jest.doMock('moment', () => {
    moment.tz.setDefault('UTC');
    return moment;
});

jest.doMock("utils/vfs_fonts", () => ({}));

window.WebSocket = jest.fn();
