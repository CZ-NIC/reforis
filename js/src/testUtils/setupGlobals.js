/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

// Fake babel (gettext) used for docs
global._ = str => str;
global.babel = {format: (str) => str};
global.ForisTranslations = {};
