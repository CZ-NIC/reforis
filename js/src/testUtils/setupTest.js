/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import mockAxios from 'jest-mock-axios';
import moment from 'moment-timezone';

// Setup axios cleanup
global.afterEach(() => {
    mockAxios.reset();
});

// Mock babel (gettext)
window._ = str => str;
window.babel = {format: (str) => str};
window.ForisTranslations = {};

// Mock scrollIntoView
window.HTMLElement.prototype.scrollIntoView = () => {
};

jest.doMock('moment', () => {
    moment.tz.setDefault('UTC');
    return moment;
});

Date.now = jest.fn(() => new Date(Date.UTC(2019, 1, 1, 12, 13, 14)).valueOf());

jest.doMock('utils/vfs_fonts', () => {
    return {};
});
