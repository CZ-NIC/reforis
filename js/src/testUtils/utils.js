/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */


export function createPortalContainer(id) {
    let portalRoot = document.getElementById(id);
        if (!portalRoot) {
            portalRoot = document.createElement('div');
            portalRoot.setAttribute('id', id);
            document.body.appendChild(portalRoot)
        }
}
