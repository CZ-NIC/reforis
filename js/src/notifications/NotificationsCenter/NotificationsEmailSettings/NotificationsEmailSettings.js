/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';

import {FORM_STATES} from '../../../forisForm/hooks';
import ForisForm from '../../../forisForm/ForisForm';

import NotificationsEmailSettingsForm from './forms/NotificationsEmailSettingsForm';
import validator from './validator';

export default function NotificationsEmailSettings({ws}) {
    const module = 'router_notifications';

    // The notifications email setting has different WS notifications action, moreover it should be updated after
    // getting update_email_settings notification.
    function notificationsEmailSettingWSLogic(setFormState, loadData) {
        ws.subscribe(module)
            .bind(module, 'update_email_settings', () => {
                setFormState(FORM_STATES.UPDATE);
                loadData()
            });
    }

    return <>
        <ForisForm
            ws={ws}
            module={module}
            prepData={prepData}
            prepDataToSubmit={prepDataToSubmit}
            validator={validator}
            anotherWSLogic={notificationsEmailSettingWSLogic}
        >
            <NotificationsEmailSettingsForm/>
        </ForisForm>
    </>
}

function prepData(formData) {
    formData = formData.emails;
    formData.common.to = formData.common.to.join(', ');
    return formData
}

function prepDataToSubmit(formData) {
    if (!formData.enabled) return {enabled: false};
    formData.common.to = formData.common.to.replace(/\s+/g, '').split(',');

    if (formData.smtp_type === 'turris')
        delete formData.smtp_custom;
    else if (formData.smtp_type === 'custom')
        delete formData.smtp_turris;
    return formData
}

