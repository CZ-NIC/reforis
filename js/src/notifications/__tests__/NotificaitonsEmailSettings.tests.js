/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';

import {act, render, getByLabelText, fireEvent} from 'react-testing-library';
import {mockedWS} from '../../testUtils/mockWS';
import mockAxios from 'jest-mock-axios';

import NotificationsEmailSettings from '../NotificationsCenter/NotificationsEmailSettings/NotificationsEmailSettings';
import {notificationsEmailSettingsFixure} from './__fixtures__/notificationsEmailSettings';


describe('<NotificationsDropdown/>', () => {
    let NotificationCenterContainer;

    beforeEach(() => {
        const mockWebSockets = new mockedWS();
        const {container} = render(<NotificationsEmailSettings ws={mockWebSockets}/>);
        mockAxios.mockResponse({data: notificationsEmailSettingsFixure()});
        NotificationCenterContainer = container
    });

    it('Enabled, smtp_type:custom', () => {
        expect(NotificationCenterContainer.firstChild).toMatchSnapshot()
    });

    it('Disabled', () => {
        act(() => {
            fireEvent.click(getByLabelText(NotificationCenterContainer, 'Enable email notifications'));
        });
        expect(NotificationCenterContainer.firstChild).toMatchSnapshot()
    });

    it('Enabled,smtp_type:turris', () => {
        act(() => {
            fireEvent.click(getByLabelText(NotificationCenterContainer, 'Turris'));
        });
        expect(NotificationCenterContainer.firstChild).toMatchSnapshot()
    })
});