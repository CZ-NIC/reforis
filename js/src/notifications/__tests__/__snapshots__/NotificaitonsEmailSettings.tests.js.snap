// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotificationsDropdown/> Disabled 1`] = `
<form>
  <h3>
    Email notifications settings
  </h3>
  <div
    class="form-group row"
  >
    <div
      class="form-label col-sm-3"
    >
      <label
        class="form-label"
        for="45"
      >
        Enable email notifications
      </label>
    </div>
    <div
      class="form-check col-sm-8"
    >
      <input
        checked=""
        id="45"
        type="checkbox"
      />
      <small
        class="form-text text-muted"
      >
        If you set SMTP provider to "Turris", the servers provided to members of the Turris project would be used. These servers do not require any additional settings. If you want to set your own SMTP server, please select "Custom" and enter required settings.
      </small>
    </div>
  </div>
  <button
    class="btn btn-primary  offset-lg-8 col-lg-3 col-sm-12"
  >
     
     
    Save
  </button>
</form>
`;

exports[`<NotificationsDropdown/> Enabled, smtp_type:custom 1`] = `
<form>
  <h3>
    Email notifications settings
  </h3>
  <div
    class="form-group row"
  >
    <div
      class="form-label col-sm-3"
    >
      <label
        class="form-label"
        for="1"
      >
        Enable email notifications
      </label>
    </div>
    <div
      class="form-check col-sm-8"
    >
      <input
        checked=""
        disabled=""
        id="1"
        type="checkbox"
      />
      <small
        class="form-text text-muted"
      >
        If you set SMTP provider to "Turris", the servers provided to members of the Turris project would be used. These servers do not require any additional settings. If you want to set your own SMTP server, please select "Custom" and enter required settings.
      </small>
    </div>
  </div>
  <div
    class="form-group row"
    id="2"
  >
    <div
      class="form-label col-sm-3"
    >
      <label
        class="form-label"
        for="2"
      >
        SMTP provider
      </label>
    </div>
    <div
      class="col-sm-8"
    >
      <div
        class="form-check form-check-inline"
      >
        <input
          class="form-check-input"
          disabled=""
          id="smtp_provider-0"
          name="smtp_provider"
          type="radio"
          value="turris"
        />
        <label
          class="form-check-label"
          for="smtp_provider-0"
        >
          Turris
        </label>
      </div>
      <div
        class="form-check form-check-inline"
      >
        <input
          checked=""
          class="form-check-input"
          disabled=""
          id="smtp_provider-1"
          name="smtp_provider"
          type="radio"
          value="custom"
        />
        <label
          class="form-check-label"
          for="smtp_provider-1"
        >
          Custom
        </label>
      </div>
      <small
        class="form-text text-muted"
      />
    </div>
  </div>
  <div
    class="form-group row"
  >
    <label
      class="form-control-label col-sm-3"
      for="3"
    >
      Recipient's email
    </label>
    <div
      class="col-sm-8"
    >
      <div
        class="input-group"
      >
        <input
          class="form-control "
          disabled=""
          id="3"
          required=""
          type="text"
          value="some@example.com"
        />
      </div>
      <div
        class="invalid-feedback"
      />
      <small
        class="form-text text-muted"
      >
        Email address of recipient. Separate multiple addresses by comma.
      </small>
    </div>
  </div>
  <div
    class="form-group row"
  >
    <label
      class="form-control-label col-sm-3"
      for="4"
    >
      Importance
    </label>
    <div
      class="col-sm-8"
    >
      <select
        class="form-control"
        disabled=""
        id="4"
      >
        <option
          value="1"
        >
          Reboot is required
        </option>
        <option
          value="2"
        >
          Reboot or attention is required
        </option>
        <option
          value="3"
        >
          Reboot or attention is required or update was installed
        </option>
      </select>
      <small
        class="form-text text-muted"
      />
    </div>
  </div>
  <div
    class="form-group row"
  >
    <div
      class="form-label col-sm-3"
    >
      <label
        class="form-label"
        for="5"
      >
        Send news
      </label>
    </div>
    <div
      class="form-check col-sm-8"
    >
      <input
        checked=""
        disabled=""
        id="5"
        type="checkbox"
      />
      <small
        class="form-text text-muted"
      >
        Send emails about new features.
      </small>
    </div>
  </div>
  <h4>
    SMTP settings
  </h4>
  <div
    class="form-group row"
  >
    <label
      class="form-control-label col-sm-3"
      for="6"
    >
      Sender address (From)
    </label>
    <div
      class="col-sm-8"
    >
      <div
        class="input-group"
      >
        <input
          class="form-control "
          disabled=""
          id="6"
          placeholder="router@example.com"
          required=""
          type="email"
          value="router@example.com"
        />
      </div>
      <div
        class="invalid-feedback"
      />
      <small
        class="form-text text-muted"
      >
        This is the address notifications are send from.
      </small>
    </div>
  </div>
  <div
    class="form-group row"
  >
    <label
      class="form-control-label col-sm-3"
      for="7"
    >
      Server address
    </label>
    <div
      class="col-sm-8"
    >
      <div
        class="input-group"
      >
        <input
          class="form-control "
          disabled=""
          id="7"
          placeholder="example.com"
          required=""
          type="text"
          value="example.com"
        />
      </div>
      <div
        class="invalid-feedback"
      />
      <small
        class="form-text text-muted"
      />
    </div>
  </div>
  <div
    class="form-group row"
  >
    <label
      class="form-control-label col-sm-3"
      for="8"
    >
      Server port
    </label>
    <div
      class="col-sm-8"
    >
      <div
        class="input-group"
      >
        <input
          class="form-control "
          disabled=""
          id="8"
          max="65535"
          min="1"
          type="number"
          value="465"
        />
      </div>
      <div
        class="invalid-feedback"
      />
      <small
        class="form-text text-muted"
      />
    </div>
  </div>
  <div
    class="form-group row"
  >
    <label
      class="form-control-label col-sm-3"
      for="9"
    >
      Security
    </label>
    <div
      class="col-sm-8"
    >
      <select
        class="form-control"
        disabled=""
        id="9"
      >
        <option
          value="none"
        >
          None
        </option>
        <option
          value="ssl"
        >
          SSL
        </option>
        <option
          value="starttls"
        >
          STARTTLS
        </option>
      </select>
      <small
        class="form-text text-muted"
      />
    </div>
  </div>
  <div
    class="form-group row"
  >
    <label
      class="form-control-label col-sm-3"
      for="10"
    >
      User
    </label>
    <div
      class="col-sm-8"
    >
      <div
        class="input-group"
      >
        <input
          class="form-control "
          disabled=""
          id="10"
          type="text"
          value=""
        />
      </div>
      <div
        class="invalid-feedback"
      />
      <small
        class="form-text text-muted"
      />
    </div>
  </div>
  <div
    class="form-group row"
  >
    <label
      class="form-control-label col-sm-3"
      for="11"
    >
      Password
    </label>
    <div
      class="col-sm-8"
    >
      <div
        class="input-group"
      >
        <input
          autocomplete="new-password"
          class="form-control "
          disabled=""
          id="11"
          type="password"
          value="test_password"
        />
        <div
          class="input-group-append"
        >
          <a
            class="input-group-text"
          >
            <i
              class="fa fa-eye"
            />
          </a>
        </div>
      </div>
      <div
        class="invalid-feedback"
      />
      <small
        class="form-text text-muted"
      />
    </div>
  </div>
  <button
    class="btn btn-primary  offset-lg-8 col-lg-3 col-sm-12"
    disabled=""
  >
    <span
      aria-hidden="true"
      class="spinner-border spinner-border-sm"
      role="status"
    />
     
     
     
    Load settings
  </button>
</form>
`;

exports[`<NotificationsDropdown/> Enabled,smtp_type:turris 1`] = `
<form>
  <h3>
    Email notifications settings
  </h3>
  <div
    class="form-group row"
  >
    <div
      class="form-label col-sm-3"
    >
      <label
        class="form-label"
        for="79"
      >
        Enable email notifications
      </label>
    </div>
    <div
      class="form-check col-sm-8"
    >
      <input
        checked=""
        id="79"
        type="checkbox"
      />
      <small
        class="form-text text-muted"
      >
        If you set SMTP provider to "Turris", the servers provided to members of the Turris project would be used. These servers do not require any additional settings. If you want to set your own SMTP server, please select "Custom" and enter required settings.
      </small>
    </div>
  </div>
  <div
    class="form-group row"
    id="80"
  >
    <div
      class="form-label col-sm-3"
    >
      <label
        class="form-label"
        for="80"
      >
        SMTP provider
      </label>
    </div>
    <div
      class="col-sm-8"
    >
      <div
        class="form-check form-check-inline"
      >
        <input
          class="form-check-input"
          id="smtp_provider-0"
          name="smtp_provider"
          type="radio"
          value="turris"
        />
        <label
          class="form-check-label"
          for="smtp_provider-0"
        >
          Turris
        </label>
      </div>
      <div
        class="form-check form-check-inline"
      >
        <input
          checked=""
          class="form-check-input"
          id="smtp_provider-1"
          name="smtp_provider"
          type="radio"
          value="custom"
        />
        <label
          class="form-check-label"
          for="smtp_provider-1"
        >
          Custom
        </label>
      </div>
      <small
        class="form-text text-muted"
      />
    </div>
  </div>
  <div
    class="form-group row"
  >
    <label
      class="form-control-label col-sm-3"
      for="81"
    >
      Recipient's email
    </label>
    <div
      class="col-sm-8"
    >
      <div
        class="input-group"
      >
        <input
          class="form-control "
          id="81"
          required=""
          type="text"
          value="some@example.com"
        />
      </div>
      <div
        class="invalid-feedback"
      />
      <small
        class="form-text text-muted"
      >
        Email address of recipient. Separate multiple addresses by comma.
      </small>
    </div>
  </div>
  <div
    class="form-group row"
  >
    <label
      class="form-control-label col-sm-3"
      for="82"
    >
      Importance
    </label>
    <div
      class="col-sm-8"
    >
      <select
        class="form-control"
        id="82"
      >
        <option
          value="1"
        >
          Reboot is required
        </option>
        <option
          value="2"
        >
          Reboot or attention is required
        </option>
        <option
          value="3"
        >
          Reboot or attention is required or update was installed
        </option>
      </select>
      <small
        class="form-text text-muted"
      />
    </div>
  </div>
  <div
    class="form-group row"
  >
    <div
      class="form-label col-sm-3"
    >
      <label
        class="form-label"
        for="83"
      >
        Send news
      </label>
    </div>
    <div
      class="form-check col-sm-8"
    >
      <input
        checked=""
        id="83"
        type="checkbox"
      />
      <small
        class="form-text text-muted"
      >
        Send emails about new features.
      </small>
    </div>
  </div>
  <div
    class="form-group row"
  >
    <label
      class="form-control-label col-sm-3"
      for="117"
    >
      Sender's name
    </label>
    <div
      class="col-sm-8"
    >
      <div
        class="input-group"
      >
        <input
          class="form-control "
          id="117"
          type="text"
          value="turris"
        />
      </div>
      <div
        class="invalid-feedback"
      />
      <small
        class="form-text text-muted"
      >
        Name of the sender - will be used as a part of the sender's email address before the "at" sign.
      </small>
    </div>
  </div>
  <button
    class="btn btn-primary  offset-lg-8 col-lg-3 col-sm-12"
  >
     
     
    Save
  </button>
</form>
`;
